lib/ProductOpener/Tags.pm

```diff
@@ -1048,176 +1148,207 @@ sub build_tags_taxonomy($$$) {
+								next if ((length($tagid2_s) > length($tagid2)) and ($tagid =~ /${tagid2_s}/));
+								next if ($tagid2_s =~ /$tagid/);
 
-							my $tagid_new = $tagid;
-							my $replaceby = "${before}${tagid2_s}${after}";
-							$tagid_new =~ s/$replace/$replaceby/e;
+								# generate the tag with substitution
+								my $tagid_new = $tagid;
+								my $replaceby = "${before}${tagid2_s}${after}";
+								# TODO: why do we need /e here ?
+								$tagid_new =~ s/$replace/$replaceby/e;
 
 
 
-							#print "computing synonyms for $tagid ($tagid0): replaceby: $replaceby - tagid4: $tagid4\n";
+								#print "computing synonyms for $tagid ($tagid0): replaceby: $replaceby - tagid4: $tagid4\n";
 
-							if (not defined $synonyms_for_extended{$tagtype}{$lc}{$tagid_c}{$tagid_new}) {
-								$synonyms_for_extended{$tagtype}{$lc}{$tagid_c}{$tagid_new} = 1;
-								$synonyms{$tagtype}{$lc}{$tagid_new} = $tagid_c;
-								if (defined $synonym_contains_synonyms{$lc}{$tagid_new}) {
-									$synonym_contains_synonyms{$lc}{$tagid_new} = clone($synonym_contains_synonyms{$lc}{$tagid});
-								}
-								else {
-									$synonym_contains_synonyms{$lc}{$tagid_new} = {};
+								if (not defined $synonyms_for_extended{$tagtype}{$lc}{$tagid_c}{$tagid_new}) {
+									# register substitution as a new synonym
+									$synonyms_for_extended{$tagtype}{$lc}{$tagid_c}{$tagid_new} = 1;
+									# register in synonyms
+									$synonyms{$tagtype}{$lc}{$tagid_new} = $tagid_c;
+									# and register the supstitution happened
+									if (defined $synonym_contains_synonyms{$lc}{$tagid_new}) {
+										# we inherit substitutions already made on original tagid
+										$synonym_contains_synonyms{$lc}{$tagid_new} = clone($synonym_contains_synonyms{$lc}{$tagid});
+									}
+									else {
+										$synonym_contains_synonyms{$lc}{$tagid_new} = {};
+									}
+									$synonym_contains_synonyms{$lc}{$tagid_new}{$tagid2_c} = 1;
+									# print STDERR "synonyms_extended : synonyms{$tagtype}{$lc}{$tagid_new} = $tagid_c (tagid: $tagid - tagid2: $tagid2 - tagid2_c: $tagid2_c - tagid2_s: $tagid2_s - replace: $replace - replaceby: $replaceby)\n";
 								}
-								$synonym_contains_synonyms{$lc}{$tagid_new}{$tagid2_c} = 1;
-								# print STDERR "synonyms_extended : synonyms{$tagtype}{$lc}{$tagid_new} = $tagid_c (tagid: $tagid - tagid2: $tagid2 - tagid2_c: $tagid2_c - tagid2_s: $tagid2_s - replace: $replace - replaceby: $replaceby)\n";
 							}
 						}
+
 					}
 
-				}
+				}  # end of substitutions on a tagid
 
-			}
+			}  # end of language code $lc loop
 
-		}
-
-		}
+		}  # on of pass loop
 
 
 		# add more synonyms: remove stopwords and deal with simple plurals
 		# -> should not be done on some taxonomies that contain only proper names
+		# TODO we could mark this kind of thing in a header for taxonomy
 		if (($tagtype ne "countries") and ($tagtype ne "origins")) {
 
+			# TODO: explan why we do it in synonyms and not in extended synonyms
```

Comment:
In fact %synonyms (used for matching) contains all keys from %synonyms_for and %synonyms_for_extended

Reply:
Ooooh this is good to note above, I will do.

Reply:
Haha, it was already explained :-D